File: codigos/problema1/analisis.py
import matplotlib
matplotlib.use('Agg')
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.gridspec as gridspec
from statsmodels.formula.api import ols
from statsmodels.stats.anova import anova_lm
from scipy.stats import f as f_dist, studentized_range
import os, warnings
warnings.filterwarnings('ignore')

# ============================================================
# PROBLEMA #1 - Factorial 4x3 - QuantEdge Capital
# ============================================================

TITULO      = "PROBLEMA #1 - Factorial: Estrategia x Regimen Volatilidad"
NOMBRE_A    = "Estrategia"
NOMBRE_B    = "Regimen"
RESPUESTA   = "Sharpe Ratio mensual"
PREFIJO_OUT = "P1"

NIVELES_A = ["Mkt.Making", "Arb.Triang", "Momentum", "Mean Rev."]
NIVELES_B = ["Baja", "Media", "Alta"]
n = 3

# Datos: para cada nivel de A, ingresar valores en orden B1*n, B2*n, B3*n
datos = [
    # A = Market Making: Baja(3), Media(3), Alta(3)
    0.77, 0.80, 0.83,   1.17, 1.20, 1.23,   1.72, 1.75, 1.78,
    # A = Arbitraje Triangular
    1.32, 1.35, 1.38,   1.82, 1.85, 1.88,   2.52, 2.55, 2.58,
    # A = Momentum
    1.07, 1.10, 1.13,   1.52, 1.55, 1.58,   2.12, 2.15, 2.18,
    # A = Mean Reversion
    1.62, 1.65, 1.68,   2.17, 2.20, 2.23,   2.97, 3.00, 3.03,
]

a = len(NIVELES_A); b = len(NIVELES_B); N = a * b * n
alpha = 0.05

col_A = [niv for niv in NIVELES_A for _ in range(b * n)]
col_B = [niv for _ in range(a) for niv in NIVELES_B for _ in range(n)]
df = pd.DataFrame({NOMBRE_A: col_A, NOMBRE_B: col_B, 'Y': datos})
df[NOMBRE_A] = pd.Categorical(df[NOMBRE_A], categories=NIVELES_A)
df[NOMBRE_B] = pd.Categorical(df[NOMBRE_B], categories=NIVELES_B)

# ============================================================
# ANOVA FACTORIAL (Tipo II)
# ============================================================
formula = f"Y ~ C({NOMBRE_A}) + C({NOMBRE_B}) + C({NOMBRE_A}):C({NOMBRE_B})"
modelo  = ols(formula, data=df).fit()
tabla_anova = anova_lm(modelo, typ=2)

print("=" * 65)
print(f"TABLA ANOVA - {TITULO}")
print("=" * 65)
print(tabla_anova.round(4))

gl_A  = a - 1; gl_B = b - 1; gl_AB = gl_A * gl_B; gl_E = a * b * (n - 1)
ft_A  = f_dist.ppf(1-alpha, gl_A,  gl_E)
ft_B  = f_dist.ppf(1-alpha, gl_B,  gl_E)
ft_AB = f_dist.ppf(1-alpha, gl_AB, gl_E)

key_A  = f"C({NOMBRE_A})"
key_B  = f"C({NOMBRE_B})"
key_AB = f"C({NOMBRE_A}):C({NOMBRE_B})"
Fc_A   = tabla_anova.loc[key_A,  'F']
Fc_B   = tabla_anova.loc[key_B,  'F']
Fc_AB  = tabla_anova.loc[key_AB, 'F']

print(f"\nFt {NOMBRE_A}  ({gl_A},{gl_E})  = {ft_A:.4f} | Fc={Fc_A:.4f} -> {'SIGNIFICATIVO' if Fc_A>ft_A else 'no sig.'}")
print(f"Ft {NOMBRE_B}  ({gl_B},{gl_E})  = {ft_B:.4f} | Fc={Fc_B:.4f} -> {'SIGNIFICATIVO' if Fc_B>ft_B else 'no sig.'}")
print(f"Ft AxB ({gl_AB},{gl_E}) = {ft_AB:.4f} | Fc={Fc_AB:.4f} -> {'SIGNIFICATIVO' if Fc_AB>ft_AB else 'no sig.'}")

# ============================================================
# TUKEY MANUAL
# ============================================================
sc_E_v  = tabla_anova.loc['Residual', 'sum_sq']
cm_E_v  = sc_E_v / gl_E
n_A     = b * n
n_B     = a * n

q_A       = studentized_range.ppf(0.95, a, gl_E)
q_B       = studentized_range.ppf(0.95, b, gl_E)
T_alpha_A = q_A * np.sqrt(cm_E_v / n_A)
T_alpha_B = q_B * np.sqrt(cm_E_v / n_B)

medias_A = df.groupby(NOMBRE_A, observed=True)['Y'].mean()
medias_B = df.groupby(NOMBRE_B, observed=True)['Y'].mean()

if Fc_A > ft_A:
    print("\n" + "=" * 65)
    print(f"TUKEY MANUAL - {NOMBRE_A}")
    print(f"  CM_E={cm_E_v:.4f}  gl_E={gl_E}  n_A={n_A}")
    print(f"  q(0.05,{a},{gl_E})={q_A:.4f}   T_alpha={T_alpha_A:.4f}")
    print("=" * 65)
    nivs_A = medias_A.sort_values(ascending=False).index.tolist()
    for i in range(len(nivs_A)):
        for j in range(i+1, len(nivs_A)):
            ni, nj = nivs_A[i], nivs_A[j]
            dif = abs(medias_A[ni] - medias_A[nj])
            sig = "SIGNIFICATIVO *" if dif > T_alpha_A else "no significativo"
            print(f"  {ni} vs {nj}: |{medias_A[ni]:.4f}-{medias_A[nj]:.4f}| = {dif:.4f} {'>' if dif>T_alpha_A else '<'} {T_alpha_A:.4f} -> {sig}")

if Fc_B > ft_B:
    print("\n" + "=" * 65)
    print(f"TUKEY MANUAL - {NOMBRE_B}")
    print(f"  CM_E={cm_E_v:.4f}  gl_E={gl_E}  n_B={n_B}")
    print(f"  q(0.05,{b},{gl_E})={q_B:.4f}   T_alpha={T_alpha_B:.4f}")
    print("=" * 65)
    nivs_B = medias_B.sort_values(ascending=False).index.tolist()
    for i in range(len(nivs_B)):
        for j in range(i+1, len(nivs_B)):
            ni, nj = nivs_B[i], nivs_B[j]
            dif = abs(medias_B[ni] - medias_B[nj])
            sig = "SIGNIFICATIVO *" if dif > T_alpha_B else "no significativo"
            print(f"  {ni} vs {nj}: |{medias_B[ni]:.4f}-{medias_B[nj]:.4f}| = {dif:.4f} {'>' if dif>T_alpha_B else '<'} {T_alpha_B:.4f} -> {sig}")

# ============================================================
# Helper: grafica de intervalos Tukey
# ============================================================
def _tukey_plot(ax, medias, T_alpha, titulo):
    niv = medias.sort_values().index.tolist()
    pares = []
    for i in range(len(niv)):
        for j in range(i+1, len(niv)):
            ni, nj = niv[i], niv[j]
            diff = float(medias[nj] - medias[ni])
            pares.append((f"{nj}-{ni}", diff, diff - T_alpha, diff + T_alpha))
    for idx, (lbl, diff, lo, hi) in enumerate(pares):
        sig = (lo > 0) or (hi < 0)
        col = 'red' if sig else 'steelblue'
        ax.plot([lo, hi], [idx, idx], color=col, linewidth=2)
        ax.plot(diff, idx, 'o', color=col, markersize=5)
    ax.set_yticks(range(len(pares)))
    ax.set_yticklabels([p[0] for p in pares], fontsize=8)
    ax.axvline(0, color='black', linewidth=0.8, linestyle='--')
    ax.set_title(titulo, fontweight='bold', fontsize=10)
    ax.set_xlabel('Diferencia de medias (IC 95%)')
    ax.grid(alpha=0.3, axis='x')

# ============================================================
# FIGURA 1: Efectos principales, interaccion y Tukey
# ============================================================
COLORES = ['#1f77b4','#ff7f0e','#2ca02c','#d62728','#9467bd','#8c564b']
media_g = df['Y'].mean()

fig = plt.figure(figsize=(18, 13))
fig.suptitle(TITULO + f"\n(Variable respuesta: {RESPUESTA})", fontsize=14, fontweight='bold')
gs = gridspec.GridSpec(2, 3, figure=fig, hspace=0.45, wspace=0.35,
                       top=0.83, bottom=0.07, left=0.07, right=0.97)

ax1 = fig.add_subplot(gs[0, 0])
mA = df.groupby(NOMBRE_A, observed=True)['Y'].mean()[NIVELES_A]
ax1.bar(NIVELES_A, mA.values, color='steelblue', edgecolor='black')
ax1.axhline(media_g, color='red', linestyle='--', linewidth=1.5, label='Media global')
ax1.set_title(f'Efecto Principal\n{NOMBRE_A}', fontweight='bold')
ax1.set_xlabel(NOMBRE_A); ax1.set_ylabel(f'Media {RESPUESTA}')
ax1.tick_params(axis='x', rotation=20)
ax1.legend(fontsize=8); ax1.grid(axis='y', alpha=0.3)

ax2 = fig.add_subplot(gs[0, 1])
mB = df.groupby(NOMBRE_B, observed=True)['Y'].mean()[NIVELES_B]
ax2.bar(NIVELES_B, mB.values, color='coral', edgecolor='black')
ax2.axhline(media_g, color='red', linestyle='--', linewidth=1.5, label='Media global')
ax2.set_title(f'Efecto Principal\n{NOMBRE_B}', fontweight='bold')
ax2.set_xlabel(NOMBRE_B); ax2.set_ylabel(f'Media {RESPUESTA}')
ax2.legend(fontsize=8); ax2.grid(axis='y', alpha=0.3)

ax3 = fig.add_subplot(gs[0, 2])
for i, nivel_b in enumerate(NIVELES_B):
    meds = df[df[NOMBRE_B] == nivel_b].groupby(NOMBRE_A, observed=True)['Y'].mean()[NIVELES_A]
    ax3.plot(NIVELES_A, meds.values, marker='o', linewidth=2,
             color=COLORES[i % len(COLORES)], label=f'{nivel_b}')
ax3.set_title(f'Grafica de Interaccion\n{NOMBRE_A} x {NOMBRE_B}', fontweight='bold')
ax3.set_xlabel(NOMBRE_A); ax3.set_ylabel(f'Media {RESPUESTA}')
ax3.tick_params(axis='x', rotation=20)
ax3.legend(fontsize=8, title=NOMBRE_B); ax3.grid(alpha=0.3)

ax4 = fig.add_subplot(gs[1, 0:2])
_tukey_plot(ax4, medias_A, T_alpha_A, f'Prueba de Tukey - {NOMBRE_A} (IC 95% manual correcto)')

ax5 = fig.add_subplot(gs[1, 2])
_tukey_plot(ax5, medias_B, T_alpha_B, f'Prueba de Tukey - {NOMBRE_B} (IC 95% manual correcto)')

BASE = os.path.dirname(os.path.abspath(__file__)) if '__file__' in dir() else r'C:\Users\ASUS\Desktop\CONTROL PC3'
SELLO_PY = "Generado en Python  |  matplotlib  |  statsmodels (ANOVA) + scipy (Tukey)"
fig.text(0.5, 0.012, SELLO_PY, ha='center', va='bottom', fontsize=13,
         style='italic', fontweight='bold', color='#1a5276',
         bbox=dict(boxstyle='round,pad=0.4', facecolor='#eaf2f8', edgecolor='#1a5276', linewidth=1.2))
ruta1 = os.path.join(BASE, 'python_1.png')
plt.savefig(ruta1, dpi=150, bbox_inches='tight'); plt.close()
print(f"\nGrafica guardada: {ruta1}")

# ============================================================
# FIGURA 2: Tabla ANOVA (imagen separada)
# ============================================================
sc_A  = tabla_anova.loc[key_A,  'sum_sq']
sc_B  = tabla_anova.loc[key_B,  'sum_sq']
sc_AB = tabla_anova.loc[key_AB, 'sum_sq']
sc_E  = sc_E_v
sc_T  = sc_A + sc_B + sc_AB + sc_E

dec_A  = 'Significativo ***' if Fc_A  > ft_A  else 'No significativo'
dec_B  = 'Significativo ***' if Fc_B  > ft_B  else 'No significativo'
dec_AB = 'Significativo ***' if Fc_AB > ft_AB else 'No significativo'

encab = ['Fuente de Variacion', 'GL', 'SC', 'CM', 'Fc', 'Ft (a=5%)', 'Conclusion']
filas = [
    [f'{NOMBRE_A} (Estrategia)', str(gl_A),  f'{sc_A:.4f}',  f'{sc_A/gl_A:.4f}',   f'{Fc_A:.4f}',  f'{ft_A:.4f}',  dec_A ],
    [f'{NOMBRE_B} (Regimen)',    str(gl_B),  f'{sc_B:.4f}',  f'{sc_B/gl_B:.4f}',   f'{Fc_B:.4f}',  f'{ft_B:.4f}',  dec_B ],
    ['Interaccion AxB',          str(gl_AB), f'{sc_AB:.4f}', f'{sc_AB/gl_AB:.4f}', f'{Fc_AB:.4f}', f'{ft_AB:.4f}', dec_AB],
    ['Error (Residual)',          str(gl_E),  f'{sc_E:.4f}',  f'{sc_E/gl_E:.4f}',   '-',            '-',            '-'   ],
    ['Total',                    str(N-1),   f'{sc_T:.4f}',  '-',                  '-',            '-',            '-'   ],
]
todas = [encab] + filas

fig2, ax_t = plt.subplots(figsize=(16, 4))
ax_t.set_xlim(0, 1); ax_t.set_ylim(0, 1); ax_t.axis('off')
fig2.suptitle(f'Tabla ANOVA - Diseño Factorial 4x3\n{TITULO}',
              fontsize=13, fontweight='bold', y=0.98)

col_x = [0.00, 0.30, 0.36, 0.48, 0.58, 0.68, 0.80, 1.00]
n_fil = len(todas); row_h = 0.82 / n_fil; y_start = 0.88
sig_map = [Fc_A > ft_A, Fc_B > ft_B, Fc_AB > ft_AB]

for i, fila in enumerate(todas):
    y_top = y_start - i * row_h; y_bot = y_top - row_h; y_mid = (y_top + y_bot) / 2
    if   i == 0:        bg, fg, fw = '#2c3e50', 'white', 'bold'
    elif 1 <= i <= 3:   bg, fg, fw = ('#d5f5e3' if sig_map[i-1] else '#ebebeb'), 'black', 'normal'
    elif i == 4:        bg, fg, fw = '#fdfefe', 'black', 'normal'
    else:               bg, fg, fw = '#eaecee', 'black', 'normal'
    for j, texto in enumerate(fila):
        x0, x1 = col_x[j], col_x[j+1]
        rect = plt.Rectangle((x0,y_bot), x1-x0, row_h, facecolor=bg,
                              edgecolor='#555555', linewidth=0.8,
                              transform=ax_t.transAxes, clip_on=True)
        ax_t.add_patch(rect)
        ax_t.text((x0+x1)/2, y_mid, texto, ha='center', va='center', fontsize=10,
                  color=fg, fontweight=fw, transform=ax_t.transAxes, clip_on=True)

fig2.text(0.5, 0.01, SELLO_PY, ha='center', va='bottom', fontsize=12,
          style='italic', fontweight='bold', color='#1a5276',
          bbox=dict(boxstyle='round,pad=0.4', facecolor='#eaf2f8', edgecolor='#1a5276', linewidth=1.2))
ruta2 = os.path.join(BASE, 'python_2.png')
plt.savefig(ruta2, dpi=150, bbox_inches='tight'); plt.close()
print(f"Tabla ANOVA guardada: {ruta2}")

File: codigos/problema1/analisis.R
# ============================================================
# PROBLEMA #1 - Factorial 4x3 - QuantEdge Capital
# Factor A = Estrategia (4 niveles) x Factor B = Regimen (3 niveles), n=3
# ============================================================
library(ggplot2); library(dplyr); library(gridExtra)

TITULO    <- "PROBLEMA #1 - Factorial: Estrategia x Regimen Volatilidad"
NOMBRE_A  <- "Estrategia"
NOMBRE_B  <- "Regimen"
RESPUESTA <- "Sharpe Ratio mensual"

niveles_A <- c("Mkt.Making", "Arb.Triang", "Momentum", "Mean Rev.")
niveles_B <- c("Baja", "Media", "Alta")
n         <- 3

datos <- c(
  # A = Market Making: Baja(3), Media(3), Alta(3)
  0.77, 0.80, 0.83,   1.17, 1.20, 1.23,   1.72, 1.75, 1.78,
  # A = Arbitraje Triangular
  1.32, 1.35, 1.38,   1.82, 1.85, 1.88,   2.52, 2.55, 2.58,
  # A = Momentum
  1.07, 1.10, 1.13,   1.52, 1.55, 1.58,   2.12, 2.15, 2.18,
  # A = Mean Reversion
  1.62, 1.65, 1.68,   2.17, 2.20, 2.23,   2.97, 3.00, 3.03
)

a <- length(niveles_A); b <- length(niveles_B); N <- a * b * n

fA <- rep(niveles_A, each = b * n)
fB <- rep(rep(niveles_B, each = n), times = a)
df <- data.frame(
  A = factor(fA, levels = niveles_A),
  B = factor(fB, levels = niveles_B),
  Y = datos
)
names(df)[1:2] <- c(NOMBRE_A, NOMBRE_B)

# ============================================================
# ANOVA
# ============================================================
formula_modelo <- as.formula(paste("Y ~", NOMBRE_A, "*", NOMBRE_B))
modelo  <- aov(formula_modelo, data = df)
resumen <- summary(modelo)[[1]]

cat(strrep("=",65),"\nTABLA ANOVA -", TITULO, "\n", strrep("=",65),"\n")
print(summary(modelo))

alpha <- 0.05
gl_A  <- a - 1; gl_B <- b - 1; gl_AB <- gl_A * gl_B; gl_E <- a * b * (n - 1)
ft_A  <- qf(1-alpha, gl_A,  gl_E)
ft_B  <- qf(1-alpha, gl_B,  gl_E)
ft_AB <- qf(1-alpha, gl_AB, gl_E)
Fc_A  <- resumen[1,"F value"]; Fc_B <- resumen[2,"F value"]; Fc_AB <- resumen[3,"F value"]

cat(sprintf("\nFt Factor A  (%d,%d) = %.4f | Fc = %.4f -> %s\n", gl_A,  gl_E, ft_A,  Fc_A,  ifelse(Fc_A >ft_A, "SIGNIFICATIVO","no significativo")))
cat(sprintf("Ft Factor B  (%d,%d) = %.4f | Fc = %.4f -> %s\n", gl_B,  gl_E, ft_B,  Fc_B,  ifelse(Fc_B >ft_B, "SIGNIFICATIVO","no significativo")))
cat(sprintf("Ft Inter AxB (%d,%d) = %.4f | Fc = %.4f -> %s\n", gl_AB, gl_E, ft_AB, Fc_AB, ifelse(Fc_AB>ft_AB,"SIGNIFICATIVO","no significativo")))

# ============================================================
# TUKEY
# ============================================================
cat("\n--- Tukey Factor A:", NOMBRE_A,"---\n")
tukey_A <- TukeyHSD(modelo, NOMBRE_A); print(tukey_A)
cat("\n--- Tukey Factor B:", NOMBRE_B,"---\n")
tukey_B <- TukeyHSD(modelo, NOMBRE_B); print(tukey_B)

# ============================================================
# GRAFICAS -> PNG  (estilo R: serif + viridis + sello)
# ============================================================
theme_R <- theme_minimal(base_family = "serif", base_size = 12) +
  theme(plot.title       = element_text(face = "bold", color = "#4b0082"),
        panel.background = element_rect(fill = "#faf7fb", color = NA),
        panel.grid.minor = element_blank(),
        legend.position  = "bottom")
PALETA_R <- c("#440154","#3b528b","#21918c","#5ec962","#fde725","#fca50a")
ACCENT_R <- "#4b0082"
SELLO_R  <- "Generado en R 4.6  |  ggplot2  |  aov + TukeyHSD"
mg <- mean(df$Y)

df_pA <- data.frame(Niv=factor(niveles_A,niveles_A), Med=tapply(df$Y,df[[NOMBRE_A]],mean))
df_pB <- data.frame(Niv=factor(niveles_B,niveles_B), Med=tapply(df$Y,df[[NOMBRE_B]],mean))

p1 <- ggplot(df_pA,aes(x=Niv,y=Med))+geom_bar(stat="identity",fill=PALETA_R[1],color="black")+
  geom_hline(yintercept=mg,color="red",linetype="dashed",linewidth=1.2)+
  labs(title=paste("Efecto Principal -",NOMBRE_A),x=NOMBRE_A,y=RESPUESTA)+theme_R+
  theme(plot.title=element_text(size=11), axis.text.x=element_text(angle=20,hjust=1))

p2 <- ggplot(df_pB,aes(x=Niv,y=Med))+geom_bar(stat="identity",fill=PALETA_R[3],color="black")+
  geom_hline(yintercept=mg,color="red",linetype="dashed",linewidth=1.2)+
  labs(title=paste("Efecto Principal -",NOMBRE_B),x=NOMBRE_B,y=RESPUESTA)+theme_R+
  theme(plot.title=element_text(size=11))

int_tab <- tapply(df$Y,list(df[[NOMBRE_B]],df[[NOMBRE_A]]),mean)
df_int  <- as.data.frame(as.table(int_tab)); names(df_int)<-c("B","A","Med")
df_int$A<-factor(df_int$A,niveles_A); df_int$B<-factor(df_int$B,niveles_B)
p3 <- ggplot(df_int,aes(x=A,y=Med,color=B,group=B))+geom_line(linewidth=1.4)+geom_point(size=3)+
  scale_color_manual(values=PALETA_R)+
  labs(title="Interaccion AxB",x=NOMBRE_A,y=RESPUESTA,color=NOMBRE_B)+theme_R+
  theme(plot.title=element_text(size=11), axis.text.x=element_text(angle=20,hjust=1))

tk_A_df <- as.data.frame(tukey_A[[NOMBRE_A]]); colnames(tk_A_df)<-c("diff","lwr","upr","p_adj"); tk_A_df$Comp<-rownames(tk_A_df); tk_A_df$Sig<-tk_A_df$p_adj<0.05
p4 <- ggplot(tk_A_df,aes(y=reorder(Comp,diff),x=diff))+
  geom_errorbarh(aes(xmin=lwr,xmax=upr,color=Sig),height=0.5,linewidth=1)+
  geom_point(aes(color=Sig),size=3)+geom_vline(xintercept=0,linetype="dashed")+
  scale_color_manual(values=c("TRUE"="#c2185b","FALSE"="#3b528b"),
                     labels=c("TRUE"="Sig.","FALSE"="No sig."),name="")+
  labs(title=paste("Tukey HSD -",NOMBRE_A),x="Dif. medias (IC 95%)",y="")+
  theme_R+theme(plot.title=element_text(size=11))

tk_B_df <- as.data.frame(tukey_B[[NOMBRE_B]]); colnames(tk_B_df)<-c("diff","lwr","upr","p_adj"); tk_B_df$Comp<-rownames(tk_B_df); tk_B_df$Sig<-tk_B_df$p_adj<0.05
p5 <- ggplot(tk_B_df,aes(y=reorder(Comp,diff),x=diff))+
  geom_errorbarh(aes(xmin=lwr,xmax=upr,color=Sig),height=0.35,linewidth=1)+
  geom_point(aes(color=Sig),size=3)+geom_vline(xintercept=0,linetype="dashed")+
  scale_color_manual(values=c("TRUE"="#c2185b","FALSE"="#21918c"),
                     labels=c("TRUE"="Sig.","FALSE"="No sig."),name="")+
  labs(title=paste("Tukey HSD -",NOMBRE_B),x="Dif. medias (IC 95%)",y="")+
  theme_R+theme(plot.title=element_text(size=11))

# Tabla ANOVA grob
sc_A<-resumen[1,"Sum Sq"]; sc_B<-resumen[2,"Sum Sq"]; sc_AB<-resumen[3,"Sum Sq"]; sc_E<-resumen[4,"Sum Sq"]
tbl_df <- data.frame(
  "Fuente"=c(NOMBRE_A,NOMBRE_B,"Interaccion AxB","Error","Total"),
  "GL"=c(gl_A,gl_B,gl_AB,gl_E,N-1),
  "SC"=round(c(sc_A,sc_B,sc_AB,sc_E,sc_A+sc_B+sc_AB+sc_E),4),
  "CM"=c(round(sc_A/gl_A,4),round(sc_B/gl_B,4),round(sc_AB/gl_AB,4),round(sc_E/gl_E,4),"-"),
  "Fc"=c(round(Fc_A,4),round(Fc_B,4),round(Fc_AB,4),"-","-"),
  "Ft"=c(round(ft_A,4),round(ft_B,4),round(ft_AB,4),"-","-"),
  "Decision"=c(ifelse(Fc_A>ft_A,"Sig ***","No sig"),ifelse(Fc_B>ft_B,"Sig ***","No sig"),
               ifelse(Fc_AB>ft_AB,"Sig ***","No sig"),"-","-"),
  check.names=FALSE, stringsAsFactors=FALSE)

anova_fills <- c(
  ifelse(Fc_A  > ft_A,  "#d9f2ee", "#efeaf5"),
  ifelse(Fc_B  > ft_B,  "#d9f2ee", "#efeaf5"),
  ifelse(Fc_AB > ft_AB, "#d9f2ee", "#efeaf5"),
  "#faf7fb", "#e8e4f0"
)
tema <- gridExtra::ttheme_default(
  core=list(bg_params=list(fill=anova_fills),fg_params=list(fontsize=9,fontfamily="serif")),
  colhead=list(bg_params=list(fill="#4b0082"),fg_params=list(col="white",fontsize=9.5,fontface="bold",fontfamily="serif")))
p_tbl <- gridExtra::arrangeGrob(
  grid::textGrob("Tabla ANOVA - Diseño Factorial 4x3", gp=grid::gpar(fontsize=11,fontface="bold",fontfamily="serif",col="#4b0082")),
  gridExtra::tableGrob(tbl_df,rows=NULL,theme=tema), ncol=1, heights=c(0.12,1))

get_script_dir <- function() {
  args <- commandArgs(trailingOnly = FALSE)
  m <- grep("^--file=", args)
  if (length(m) > 0) return(dirname(normalizePath(sub("^--file=", "", args[m]))))
  for (i in sys.nframe():1) {
    of <- sys.frame(i)$ofile
    if (!is.null(of)) return(dirname(normalizePath(of)))
  }
  if (requireNamespace("rstudioapi", quietly = TRUE) && rstudioapi::isAvailable()) {
    pth <- rstudioapi::getSourceEditorContext()$path
    if (!is.null(pth) && nzchar(pth)) return(dirname(normalizePath(pth)))
  }
  getwd()
}
base_dir <- get_script_dir()
ruta   <- file.path(base_dir, "r_1.png")
ruta_t <- file.path(base_dir, "r_2.png")

titulo_g <- grid::textGrob(TITULO, gp=grid::gpar(fontsize=12,fontface="bold",fontfamily="serif",col="#4b0082"))
sello_g  <- grid::textGrob(SELLO_R, gp=grid::gpar(fontsize=11,fontface="italic",fontfamily="serif",col=ACCENT_R))

# Paneles reordenados (Tukey arriba) para diferenciar del layout de Python
png(ruta, width=1600, height=1100, res=130)
print(grid.arrange(p1,p2,p3,p4,p5, layout_matrix=rbind(c(4,4,5),c(1,2,3)), top=titulo_g, bottom=sello_g))
dev.off()
cat("\nGrafica guardada:", ruta, "\n")

png(ruta_t, width=1300, height=470, res=130)
print(grid.arrange(p_tbl, bottom=sello_g))
dev.off()
cat("Tabla  guardada:", ruta_t, "\n")
